### ISSUE TITLE
about training details

### ISSUE THREAD

**Issue by EliotChenKJ** [temporal-cycle-consistency/temporal-cycle-consistency.github.io#1]
hi,
    l have to say, your differentiate TCC idea is brilliant. l've already read ur TCC paper, but i have some question about training details, hope u can help me get through it.

    1. In the paper, the training requires a pair of sequences to achieve the alignment work. The hyperparameters table show that your number of frames = 20, so this "number of frames" means that each encoding requires K frames to put into the encoder? 

    2. To train the network, you randomly select frames example from two sequences until the loss converge. But if there are too many frames in training sequences pair to compare distance, won't it require a lot of GPU memory?

   Looking forward to get ur answer.

**Comment by debidatta** [comment]
Glad that you liked the paper. 

Sorry for the confusion. Following are the answers to your questions:
1. Number of frames refers to how many time-steps of one video are selected in a batch. Since videos are of variable length and can be longer than what might fit in GPU memory, we randomly sample 20 time-steps of every video. At each time-step, we have the frame at that time step and the past (K-1) context time-steps. We use a small value of K=2, but increase the stride at which context frames are calculated to ensure the embeddings still have decent context. The encoder takes these 20 * K frames as input and produces 20 embeddings (one at each sampled timestamp).

2. Yes, if there are too many frames in the video then you cannot use all the timestamps to calculate the embeddings due to GPU memory constraints. For the datasets we considered, randomly sampling a fixed number of timesteps(20) from the full videos worked fine but this heuristic might not work for other datasets. The usual tricks of using smaller resolution, shallower/thinner convolutional feature extractors might help to handle longer videos.

**Comment by EliotChenKJ** [comment]
OK，it is very specific and i understand now. Thank for your answer~

**Comment by EliotChenKJ** [comment]
Sorry for the following-up question, in the cycle-back classification / regression design, why does the **cross-entropy loss** serve worse than **cycle-back regression loss**? I think that using the [0,0,0,1,0...] label to constrain the cycle-back loss is much more powerful than [0.01, 0.02, 0.4, 1, ...](just a scratch) gaussian-like label, isn't it? 
Maybe i consider it in a wrong way, hope u can help me with this.

**Comment by EliotChenKJ** [comment]
wops~It turns out that your classification loss uses CE loss rather than BCE loss. Then why not use BCE loss~

**Comment by debidatta** [comment]
No worries. Happy to clarify issues.

There are two reasons why the cycle-back regression loss is better than the cycle-back classification loss:
1) The regression loss is aware of the timestamps of the frames and is able to penalize based on how **severe** the mistake was. For example, if the ground truth frame is 10 and after cycling back the method predicts frame 9, then regression loss penalizes this less than the situation where the method predicts frame 1. For the classification loss both these errors are the same as 1 and 9 are **different** from 10. In videos neighboring frames look quite similar, so it helps to model the temporally coherent nature of videos by considering the timestamps in the loss as well.

2) The regression loss can adapt to the speed at which the action is happening. People might be doing the same action slowly or quickly in different videos (or even different parts of the same video).  If someone is doing an action slowly then the neighboring frames all look quite similar and the distribution after cycling-back shouldn't be too peaky. If the action is being done rather quickly, the neighboring frames look quite different and the distribution can be peaky. We can use the probability distribution (beta) while cycling-back to model this speed difference. 

We estimate the variance of the beta distribution to decide how much we should penalize the predictions. If the frames look different then the distribution is peaky.  If we make a small mistake but with a peaky distribution (fast action) then the penalty is high. If frames look similar (slow action), the beta distribution has a plateau. In that case we penalize small mistakes less. 

It is possible to use other kinds of losses to impose the cycle-consistency constraint. Performance will depend on nature of the actions in the dataset.

Hope this helped.

**Comment by EliotChenKJ** [comment]
Oh, the distance between two embedding of the cycle-back process and the speed difference did matter!
    Thank you for your answers~
